cpp/src/arrow/gpu/cuda_context.h

```diff
@@ -140,6 +142,71 @@ class ARROW_EXPORT CudaDevice : public Device {
   /// \param[in] size The buffer size in bytes
   Result<std::shared_ptr<CudaHostBuffer>> AllocateHostBuffer(int64_t size);
 
+  /// \brief EXPERIMENTAL: Wrapper for CUstreams
+  ///
+  /// Does not *own* the CUstream object which must be separately constructed
+  /// and freed using cuStreamCreate and cuStreamDestroy (or equivalent).
+  /// Default construction will use the cuda default stream, and does not allow
+  /// construction from literal 0 or nullptr.
+  class ARROW_EXPORT Stream : public Device::Stream {
+   public:
+    explicit Stream(std::shared_ptr<CudaContext> ctx) noexcept
+        : context_{std::move(ctx)}, stream_{} {}
+
+    ~Stream() = default;
+    explicit Stream(std::shared_ptr<CudaContext> ctx, CUstream stream) noexcept
+        : context_{std::move(ctx)}, stream_{stream} {}
```

Comment:
It would be nice to have a constructor that just works on the default CUDA context similar to how `MakeDeviceSyncEvent` and `WrapDeviceSyncEvent` work. Should we follow the same pattern for Streams here?

Reply:
I think I missed something: why does Stream need to own instead of just wrap? It seems that stream lifetime management will *always* be handled elsewhere. For example, say there's an application built with CUDA which is adding an arrow integration: the application will already have a pool of CUstreams which only need to be wrapped in arrow Device::Streams when calling an arrow function. I don't think an arrow function will ever take ownership of the stream away from the hypothetical application, so to me it seems we don't need arrow functions to wrap them in smart pointers for dynamic lifetimes or functions for producing new streams (MakeStream).

To put this another way: would we ever need to produce a `vector<shared_ptr<Stream>>` where the wrapped streams are a mix of `CUstream` and `hipStream_t`? Even if an application were using both ROCM and CUDA at the same time, the streams would (I'm certain) be maintained in separate pools which obviates the need for polymorphic lifetime management.

Reply:
I tend to agree with both of you and lean towards the synchronicity and convenience side of things. It's nice having the APIs be similar between the events and streams and so I lean more towards the ownership of the stream itself. 

That said, I'm not opposed to having it just wrap and not own. What do you think @bkietz?

Reply:
I think the potential aggravation due to arrow's failure to integrate with a user's existing stream pool is higher than the potential due to a user needing to implement their own stream pool.

Reply:
What's the issue with someone using `Stream::WrapStream` and passing `nullptr` if they don't want any lifetime management or handling something like a shared pointer in a release function if desired?

Reply:
Yea, I'd agree with @kkraus14 that if they are trying to integrate seamlessly with an existing stream pool they could use `Device::WrapStream` and pass `nullptr` or whatever other releasing back to the pool they want in the release function.

Reply:
There's not a fundamental issue, it's just an IMHO: avoiding ownership entirely here will be the more user-friendly API. Certainly using `Stream::WrapStream` as you describe will also work